src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,269 @@
+"""
+    SSymmetricCompact{N, T, L} <: StaticMatrix{N, N, T}
+
+A [StaticArray](@ref) subtype that represents a symmetric matrix. Unlike
+`LinearAlgebra.Symmetric`, `SSymmetricCompact` stores only the lower triangle
+of the matrix (as an `SVector`). The lower triangle is stored in column-major order.
+For example, for an `SSymmetricCompact{3}`, the indices of the stored elements
+can be visualized as follows:
+
+```
+┌ 1 ⋅ ⋅ ┐
+| 2 4 ⋅ |
+└ 3 5 6 ┘
+```
+
+Type parameters:
+* `N`: matrix dimension;
+* `T`: element type for lower triangle;
+* `L`: length of the `SVector` storing the lower triangular elements.
+
+Note that `L` is always the `N`th [triangular number](https://en.wikipedia.org/wiki/Triangular_number).
+
+An `SSymmetricCompact` may be constructed either:
+
+* from an `AbstractVector` containing the lower triangular elements; or
+* from a `Tuple` containing both upper and lower triangular elements (in column major order); or
```

Comment:
So I guess this inconsistency between `Tuple` and `AbstractVector` is due to the rest of `StaticArrays` assuming dense storage? That's unfortunate.

Does this `Tuple` constructor ignoring the top half, combined with the `similar_type` machinery mean that a lot of `StaticArrays` functionality is actually broken with `SSymmetricCompact`? For example, matrix multiply of `SSymmetricCompact` with `SMatrix`?

Reply:
It's a bit of a pickle (see https://github.com/JuliaArrays/StaticArrays.jl/pull/358#issue-162781070, item 2), but I think the current implementation still makes sense.

Any `StaticArray` definitely needs to be constructable from a `Tuple`, which needs to have the interpretation of the full list of column-major-ordered elements.; that's the basic assumption in `convert.jl`. That also means we can't use `Tuple` as the type of the `lowertriangle` field without causing all kinds of confusion with the constructors. So I chose to use an `SVector` as the storage type for the lower triangular elements and designed the constructors in accordance. The precedent for this was `SDiagonal` (which has since been replaced with `Symmetric{T, SVector{<:Any, T}`; maybe the same will happen for this case in time). 

Note that the `StaticMatrix` subtype constructors / `convert` methods currently also have the job of implementing `reshape`,

https://github.com/JuliaArrays/StaticArrays.jl/blob/407c65fb2ae652a439d09d82edf1a22eaab58dc6/src/convert.jl#L9-L10

which I'm not 100% on board with anyway. `SSymmetricCompact` hijacks this particular use case, but I'd say it's niche enough that it's OK. Alternatively, a new type could be added precisely for this purpose, but I don't think there's much of an advantage in doing so.

> Does this Tuple constructor ignoring the top half, combined with the similar_type machinery mean that a lot of StaticArrays functionality is actually broken with SSymmetricCompact? For example, matrix multiply of SSymmetricCompact with SMatrix?

No. `similar_type` is not overloaded for `SSymmetricCompact`, so the default `SMatrix` is used. So in that sense, `SSymetricCompact` is very similar to e.g. `Rotations.Quat`, and so the only thing that really matters is that `getindex(::StaticArray, i::Int)` works.